gossipsub.go

```diff
@@ -208,17 +371,34 @@ func (gs *GossipSubRouter) handleIHave(p peer.ID, ctl *pb.ControlMessage) []*pb.
 		iwantlst = append(iwantlst, mid)
 	}
 
+	// ask in random order
+	shuffleStrings(iwantlst)
+
 	return []*pb.ControlIWant{&pb.ControlIWant{MessageIDs: iwantlst}}
 }
 
 func (gs *GossipSubRouter) handleIWant(p peer.ID, ctl *pb.ControlMessage) []*pb.Message {
+	// we don't respond to IWANT requests from any peer whose score is below the gossip threshold
+	score := gs.score.Score(p)
+	if score < gs.gossipThreshold {
+		log.Debugf("IWANT: ignorin peer %s with score below threshold [score = %f]", p, score)
+		return nil
+	}
+
 	ihave := make(map[string]*pb.Message)
 	for _, iwant := range ctl.GetIwant() {
 		for _, mid := range iwant.GetMessageIDs() {
-			msg, ok := gs.mcache.Get(mid)
-			if ok {
-				ihave[mid] = msg
+			msg, count, ok := gs.mcache.GetForPeer(mid, p)
```

Comment:
This creates an entry in the mcache for each message ID the peer wants. Can't a peer just spoof random message IDs, and send us huge IWANTs up to the max message size, to make us OOM?

Reply:
One way to prevent this is to track the IHAVEs we've sent, and interesect the incoming IWANTs against them. If the peer is asking for random stuff, we need to decrement their score or blacklist them entirely.

Reply:
It will only create an entry _iff_ we have the message in mcache, and it will last for just a few seconds (until we delete the message from the cache).
I don't think it's a plausible attack vector.